src/Membership.sol

```diff
@@ -0,0 +1,427 @@
+
+    /// @dev contract initializer
+    /// @param _priceCalculator Address of an instance of IPriceCalculator
+    /// @param _maxTotalRateLimitPerEpoch Maximum total rate limit of all memberships in the tree
+    /// @param _minRateLimitPerMembership Minimum rate limit of one membership
+    /// @param _maxRateLimitPerMembership Maximum rate limit of one membership
+    /// @param _expirationTerm Membership expiration term
+    /// @param _gracePeriod Membership grace period
+    function __Membership_init(
+        address _priceCalculator,
+        uint32 _maxTotalRateLimitPerEpoch,
+        uint32 _minRateLimitPerMembership,
+        uint32 _maxRateLimitPerMembership,
+        uint32 _expirationTerm,
+        uint32 _gracePeriod
+    )
+        internal
+    {
+        priceCalculator = IPriceCalculator(_priceCalculator);
+        maxTotalRateLimitPerEpoch = _maxTotalRateLimitPerEpoch;
+        maxRateLimitPerMembership = _maxRateLimitPerMembership;
+        minRateLimitPerMembership = _minRateLimitPerMembership;
+        billingPeriod = _expirationTerm;
+        gracePeriod = _gracePeriod;
+    }
+
+    /// @notice Checks if a user message limit is valid. This does not take into account whether we the total
+    /// memberships have reached already the `maxTotalRateLimitPerEpoch`
+    /// @param userMessageLimit The user message limit
+    /// @return true if the user message limit is valid, false otherwise
+    function isValidUserMessageLimit(uint32 userMessageLimit) external view returns (bool) {
+        return userMessageLimit >= minRateLimitPerMembership && userMessageLimit <= maxRateLimitPerMembership;
+    }
+
+    /// @dev acquire a membership and trasnfer the fees to the contract
+    /// @param _sender address of the owner of the new membership
+    /// @param _idCommitment the idcommitment of the new membership
+    /// @param _rateLimit the user message limit
+    /// @param _numberOfPeriods the number of periods the user wants to acquire
+    /// @return index the index in the merkle tree
+    /// @return reusedIndex indicates whether a new leaf is being used or if using an existing leaf in the merkle tree
+    function _acquireMembership(
+        address _sender,
+        uint256 _idCommitment,
+        uint32 _rateLimit,
+        uint32 _numberOfPeriods
+    )
+        internal
+        returns (uint32 index, bool reusedIndex)
+    {
+        if (_numberOfPeriods == 0) revert NumberOfPeriodsCantBeZero();
+
+        (address token, uint256 amount) = priceCalculator.calculate(_rateLimit, _numberOfPeriods);
+        (index, reusedIndex) =
+            _setupMembershipDetails(_sender, _idCommitment, _rateLimit, _numberOfPeriods, token, amount);
+        _transferFees(_sender, token, amount);
+    }
+
+    function _transferFees(address _from, address _token, uint256 _amount) internal {
+        if (_token == address(0)) {
```

Comment:
Contract accepts ETH too. For this the token to set in the price calculator is `address(0)`.

Reply:
Shouldn't the contract call revert if ETH is being sent to it? The spec only mentions DAI as the accepted token. Also, for accepting ETH we'd need to do price conversion from USD to ETH, for which we'd need an oracle. I'd say that for the first implementation accepting ETH should be forbidden.

Reply:
I added the functionality to make the contract future proof, and would be up to us to determine whether we want to use ETH or not, but since there's no immediate plans for this, I'll drop that part of the code, specially since it would mean saving few gas due to not have to do the comparison against address(0). 
We can always add this back via an upgrade

Reply:
Support for ETH removed in https://github.com/waku-org/waku-rlnv2-contract/pull/13/commits/7f60c538b320a50e12bf0c47fd3ccf0fc81e659e